### ISSUE TITLE
Multitenant GW 3.2.0+ waiting forever to read a message from client GW

### ISSUE THREAD

**Issue by Alfiva** [universAAL/remote#450]
_Originally Opened: @Alfiva (2014-11-21 11:46:51_)
_Originally Closed: 2015-07-31 13:20:44_

This bug was the actual issue behind the original issue #448.

The problem happens with Multitenant GW 3.2.0 and above, BUT NOT WITH 3.1.2-SNAPSHOT.

What happens is that, in an usual initial communication (registering uaal members) between server GW and client GW, these messages in the serer are received correctly: 

Read message ConnectRequest going to handle it
Read message HighPush going to handle it

But the next message is stuck in AbstractSocketCommunicationHandler.readMessage (line 66) forever. It happens because in  Serializer.unmarshalMessage line 141 it reads bytes from the socket, and the read method blocks until bytes are available, which are not.

I am going to compare 3.1.2-S and 3.2.0 to see what is different, but if someone has any idea first please post it. I need to fix this today (or else I will use GW 3.1.2-S with uAAL 3.2.0)

--

From: _this issue has been automatically imported from our old issue tracker_

**Comment by Alfiva** [comment]
_Original author: @Alfiva (2014-11-21 13:23:34)_

finally found out what is going on: It is not a problem with the versions, 3.2.1-S, 3.2.0 and 3.2.1-S all behave the same (they are in fact the same code).

The problem appears when all bundles (the GW and the apps whose services are imp/exported) start together &quot;at the same time&quot;, that is, a normal execution of Karaf with all bundles already installed. So, there must be some race condition somewhere...

If you install/restart the applications *after* the MW and the GW have &quot;settled&quot; it works OK. For now I will continue to work with the GW in this way (manually restarting bundles) but for deployment this must be solved.

**Comment by Alfiva** [comment]
_Original author: @Alfiva (2015-07-31 11:22:26)_

I am hunting down the bug and I think I found it - this is just for keeping track:

At least in CLIENT mode, the problem is with the cycle beforeRun &gt; loopRun &gt; afterRun in AbstractLinkHandler.

The scan of existing uaal bus members is made in beforeRun. When an existing bus member is detected it tries to export it, but it will never receive a response from the Server, because it has not yet reached loopRun, which is the one enabling the reception of messages from the server.

The consequence seems to be that the registration of (existing) bus members during beforeRun fails. Then goes into loopRun and works as usual, correctly exporting bus members AFTER that point.

The solution should be to have the exporting of existing bus members happen AFTER loopRun is reached, but I have to be careful not to skip phases...

**Comment by Alfiva** [comment]
_Original author: @Alfiva (2015-07-31 13:20:44)_

I think I fixed this by putting the process of exporting existing members in a separate thread. This way when it blocks the GW continues and reaches loopRun, and will eventually get the response (or loopRun actually gets run before the export thread...).

However this was not enough to fix the issue, since the Server side also had problems and did not export its own existing members at startup. After debugging I found that in ServerSocketCommHandler there was this:

mySession.setStatus(SessionEvent.SessionStatus.CONNECTED);
gw.newSession(socket.toString(), mySession);

Where newSession makes the GW register for Session changes, AFTER the CONNECTED session change, and thus it missed the trigger to export the existing members.

I fixed this by extracting the s.addSessionEventListener registration from the newSession method and put it before the CONNECTED change.

Now it seems everything is properly exported in both directions even if members were started before the GW connection was made.
